pkg/proxy/iptables/proxier.go

```diff
@@ -959,12 +959,17 @@ func (proxier *Proxier) syncProxyRules() {
 			"-d", fmt.Sprintf("%s/32", svcInfo.clusterIP.String()),
 			"--dport", fmt.Sprintf("%d", svcInfo.port),
 		}
-		if proxier.masqueradeAll {
+
+		// There is no need to masquerade intra cluster traffic (unless forced to)
+		// However, if no clusterCIDR is specified we cannot distinguish
+		// intra cluster traffic from other traffic --> masquerade all
+		if proxier.masqueradeAll || len(proxier.clusterCIDR) == 0 {
 			writeLine(natRules, append(args, "-j", string(KubeMarkMasqChain))...)
-		}
-		if len(proxier.clusterCIDR) > 0 {
+		} else {
 			writeLine(natRules, append(args, "! -s", proxier.clusterCIDR, "-j", string(KubeMarkMasqChain))...)
 		}
+		// External (or all) traffic to svc cluster is Masqueraded
```

Comment:
What happens when packet from outside hits the clusterIP ?
In absence of a masquerade rule, it won;t be able to get back.

This is what happened with issue #36835.
The setup in the issue is *not* gke.
When clusterIP was specified, the masquerade was added and it solved the problem.

We want a masquerade even in the absence of clusterIP to account for the fact that traffic may be coming from outside.